examples/OpticalFlow/flownet2s.py

```diff
@@ -0,0 +1,224 @@
+but to rely on tensorpack's "OfflinePredictor" for easier inference.
+
+The ported pre-trained Caffe-model is here
+http://files.patwie.com/recipes/models/flownet2-s.npz
+
+It has the original license:
+
+```
+    Pre-trained weights are provided for research purposes only and without any warranty.
+
+    Any commercial use of the pre-trained weights requires FlowNet2 authors consent.
+    When using the the pre-trained weights in your research work, please cite the following paper:
+
+    @InProceedings{IMKDB17,
+      author       = "E. Ilg and N. Mayer and T. Saikia and M. Keuper and A. Dosovitskiy and T. Brox",
+      title        = "FlowNet 2.0: Evolution of Optical Flow Estimation with Deep Networks",
+      booktitle    = "IEEE Conference on Computer Vision and Pattern Recognition (CVPR)",
+      month        = "Jul",
+      year         = "2017",
+      url          = "http://lmb.informatik.uni-freiburg.de//Publications/2017/IMKDB17"
+    }
+```
+
+To verify this against the Caffe-model, please use the test-case:
+
+    http://files.patwie.com/recipes/models/flownet2-s-expected-values.npy (containing item 07446 of flying-chairs)
+
+and run
+
+    python flownet2s.py --gpu 0 --verify --load flownet2-s.npz --expect flownet2-s-expected-values.npy
+
+To run it on actual data:
+
+    python flownet2s.py --gpu 0 \
+        --left left_img.ppm \
+        --right right_img.ppm \
+        --apply --load flownet2-s.npz
+
+"""
+
+BATCH_SIZE = 16
+SHAPE = None
+CHANNELS = 3
+
+
+def pad(x, p=3):
+    """Pad tensor in H, W
+
+    Remarks:
+        TensorFlow uses "ceil(input_spatial_shape[i] / strides[i])" rather than explicit padding
+        like Caffe, pyTorch does. Hence, we need to pad here beforehand.
+
+    Args:
+        x (tf.tensor): incoming tensor
+        p (int, optional): padding for H, W
+
+    Returns:
+        tf.tensor: padded tensor
+    """
+    return tf.pad(x, [[0, 0], [0, 0], [p, p], [p, p]])
+
+
+def resize(x, factor=4):
+    """Resize input tensor with unkown input-shape by a factor
+
+    Args:
+        x (tf.Tensor): tensor NCHW
+        factor (int, optional): resize factor for H, W
+
+    Returns:
+        tf.Tensor: resized tensor NCHW
+    """
+    shp = tf.shape(x)[2:] * factor
+    shp = tf.Print(shp, [shp])
+    # NCHW -> NHWC
+    x = tf.transpose(x, [0, 2, 3, 1])
+    x = tf.image.resize_bilinear(x, shp, align_corners=True)
+    # NHWC -> NCHW
+    return tf.transpose(x, [0, 3, 1, 2])
+
+
+class FlowNet2S(ModelDesc):
+    def inputs(self):
+        return [tf.placeholder(tf.float32, (None, CHANNELS, SHAPE, SHAPE), 'left'),
+                tf.placeholder(tf.float32, (None, CHANNELS, SHAPE, SHAPE), 'right')]
+
+    def graph_structure(self, left, right):
+        # this is refactored into a separate function for a potential re-implementation of the entire FlowNet2
+        x = tf.stack([left, right], axis=2)
+        rgb_mean = tf.reduce_mean(x, axis=[0, 2, 3, 4], keep_dims=True)
+        x = (x - rgb_mean) / 255.
```

Comment:
Perhaps it does not matter much, but is this actually what they're doing? It seems to be some custom caffe code so I didn't follow much into it.

Reply:
Channel-wise rgb-mean subtraction for each input (note NCHW --> NTCHW reduce over NCHW).